components/board/genericlinux/i2c.go

```diff
@@ -0,0 +1,125 @@
+package genericlinux
+
+import (
+	"context"
+	"encoding/binary"
+	"sync"
+
+	"github.com/edaniels/golog"
+	"periph.io/x/conn/v3/i2c"
+	"periph.io/x/conn/v3/i2c/i2creg"
+	"periph.io/x/host/v3"
+
+	"go.viam.com/rdk/components/board"
+)
+
+func init() {
+	if _, err := host.Init(); err != nil {
+		golog.Global().Debugw("error initializing host", "error", err)
+	}
+}
+
+type i2cBus struct {
+	internal i2c.BusCloser
+	mu       sync.Mutex
+}
+
+func newI2cBus(deviceName string) (*i2cBus, error) {
+	// We return a pointer to an i2cBus instead of an i2cBus itself so that we can return nil if
+	// something goes wrong.
+	bus, err := i2creg.Open(deviceName)
+	if err != nil {
+		return nil, err
+	}
+	return &i2cBus{internal: bus}, nil
+}
+
+// This lets the i2cBus type implement the board.I2C interface.
+func (bus *i2cBus) OpenHandle(addr byte) (board.I2CHandle, error) {
+	bus.mu.Lock() // Lock the bus so no other handle can use it until this one is closed.
+	return &i2cHandle{device: &i2c.Dev{Bus: bus.internal, Addr: uint16(addr)}, mu: &bus.mu}, nil
+}
+
+type i2cHandle struct {
+	device *i2c.Dev    // Will become nil if we close the connection
+	mu     *sync.Mutex // Points to the I2C bus' mutex
+}
+
+// This helps the i2cHandle struct implement the board.I2CHandle interface.
+func (h *i2cHandle) Write(ctx context.Context, tx []byte) error {
+	return h.device.Tx(tx, nil)
+}
+
+// This helps the i2cHandle struct implement the board.I2CHandle interface.
+func (h *i2cHandle) Read(ctx context.Context, count int) ([]byte, error) {
+	buffer := make([]byte, count)
+	err := h.device.Tx(nil, buffer)
+	if err != nil {
+		return nil, err
+	}
+	return buffer, nil
+}
+
+// This helps the i2cHandle struct implement the board.I2CHandle interface.
+func (h *i2cHandle) ReadByteData(ctx context.Context, register byte) (byte, error) {
+	buffer, err := h.Read(ctx, 1)
+	if err != nil {
+		return 0, err
+	}
+	return buffer[0], nil
+}
+
+// This is a private helper function, used to implement the rest of the board.I2CHandle interface.
+func (h *i2cHandle) transactAtRegister(register byte, w, r []byte) error {
+	if w == nil {
+		w = []byte{}
+	}
+	fullW := make([]byte, len(w)+1)
+	fullW[0] = register
+	copy(fullW[1:], w)
+	return h.device.Tx(fullW, r)
+}
+
+// This helps the i2cHandle struct implement the board.I2CHandle interface.
+func (h *i2cHandle) WriteByteData(ctx context.Context, register, data byte) error {
+	return h.transactAtRegister(register, []byte{data}, nil)
+}
+
+// This helps the i2cHandle struct implement the board.I2CHandle interface.
+func (h *i2cHandle) ReadWordData(ctx context.Context, register byte) (uint16, error) {
+	result := make([]byte, 2)
+	err := h.transactAtRegister(register, nil, result)
+	if err != nil {
+		return 0, err
+	}
+	return binary.BigEndian.Uint16(result), nil
```

Comment:
The only mentions of "MSB" in that link are talking about the most significant _bit_ when describing the address of the device you want to talk to. When you read two adjacent registers from a single device and want to treat them as a single word, which one should go first? The ADXL345 was little endian, but the MPU6050 was big endian. (For neither one did I read things word-by-word: I read 6+ bytes at a time to get X, Y, and Z data all in the same call, and split up the bytes as relevant on my own.)

TBH, I question whether these functions should exist in the interface at all: I2C devices could store whatever data they want wherever they want to, and it's up to the caller to interpret the byte array that they get back. When I was using https://github.com/d2r2/go-i2c, it had both BE and LE implementations of the 16-bit operations, because different devices need different endiannesses.

Reply:
Ah okay, then you answered my question - if we can just let the driver handle endianness. Then, I agree, we should not worry about it on board at all, leave the interpretation to the driver sending over that data, and we handle the most general storing capabilities on board.

Reply:
I've made https://github.com/viamrobotics/rdk/pull/1854 to clean this up.